File: uploader/src/main/java/net/hearthstats/ui/MatchPanel.java
package net.hearthstats.ui;

import static net.hearthstats.util.Translations.t;

import java.awt.Color;
import java.awt.Desktop;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.net.URI;

import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

import net.hearthstats.Config;
import net.hearthstats.Constants;
import net.hearthstats.HearthstoneMatch;
import net.hearthstats.Main;
import net.hearthstats.analysis.HearthstoneAnalyser;
import net.miginfocom.swing.MigLayout;


public class MatchPanel extends JPanel {
  private JTextField _currentOpponentNameField;
  private JLabel _currentMatchLabel;
  private JCheckBox _currentGameCoinField;
  private JTextArea _currentNotesField;
  private JButton _lastMatchButton;
  private HearthstoneMatch _lastMatch;

  public void setLastMatch(HearthstoneMatch _lastMatch) {
    this._lastMatch = _lastMatch;
  }

  private JComboBox _currentOpponentClassSelect;
  private JComboBox _currentYourClassSelector;

  private Boolean _currentMatchEnabled = false;

  public MatchPanel() {

    MigLayout layout = new MigLayout();
    setLayout(layout);

    // match label
    add(new JLabel(" "), "wrap");
    _currentMatchLabel = new JLabel();
    add(_currentMatchLabel, "skip,span,wrap");

    add(new JLabel(" "), "wrap");

    String[] localizedClassOptions = new String[Constants.hsClassOptions.length];
    localizedClassOptions[0] = "- " + t("undetected") + " -";
    for (int i = 1; i < localizedClassOptions.length; i++)
      localizedClassOptions[i] = t(Constants.hsClassOptions[i]);

    // your class
    add(new JLabel(t("match.label.your_class") + " "), "skip,right");
    _currentYourClassSelector = new JComboBox<>(localizedClassOptions);
    add(_currentYourClassSelector, "wrap");

    // opponent class
    add(new JLabel(t("match.label.opponents_class") + " "), "skip,right");
    _currentOpponentClassSelect = new JComboBox<>(localizedClassOptions);
    add(_currentOpponentClassSelect, "wrap");

    // Opponent name
    add(new JLabel("Opponent's Name: "), "skip,right");
    _currentOpponentNameField = new JTextField();
    _currentOpponentNameField.setMinimumSize(new Dimension(100, 1));
    _currentOpponentNameField.addKeyListener(new KeyAdapter() {
      public void keyReleased(KeyEvent e) {
        HearthstoneAnalyser.hsMatch().opponentName_$eq(
            _currentOpponentNameField.getText().replaceAll("(\r\n|\n)", "<br/>"));
      }
    });
    add(_currentOpponentNameField, "wrap");

    // coin
    add(new JLabel(t("match.label.coin") + " "), "skip,right");
    _currentGameCoinField = new JCheckBox(t("match.coin"));
    _currentGameCoinField.setSelected(Config.showHsClosedNotification());
    _currentGameCoinField.addChangeListener(new ChangeListener() {
      public void stateChanged(ChangeEvent e) {
        HearthstoneAnalyser.hsMatch().coin_$eq(_currentGameCoinField.isSelected());
      }
    });
    add(_currentGameCoinField, "wrap");

    // notes
    add(new JLabel(t("match.label.notes") + " "), "skip,wrap");
    _currentNotesField = new JTextArea();
    _currentNotesField.setBorder(BorderFactory.createCompoundBorder(
        BorderFactory.createMatteBorder(1, 1, 1, 1, Color.black),
        BorderFactory.createEmptyBorder(3, 6, 3, 6)));
    _currentNotesField.setMinimumSize(new Dimension(350, 150));
    _currentNotesField.setBackground(Color.WHITE);
    _currentNotesField.addKeyListener(new KeyAdapter() {
      public void keyReleased(KeyEvent e) {
        HearthstoneAnalyser.hsMatch().notes_$eq(_currentNotesField.getText());
      }
    });
    add(_currentNotesField, "skip,span");

    add(new JLabel(" "), "wrap");

    // last match
    add(new JLabel(t("match.label.previous_match") + " "), "skip,wrap");
    _lastMatchButton = new JButton("[n/a]");
    _lastMatchButton.addActionListener(new ActionListener() {
      @Override
      public void actionPerformed(ActionEvent arg0) {
        String url = "Arena".equals(_lastMatch.mode()) ? "http://hearthstats.net/arenas/new"
            : _lastMatch.editUrl();
        try {
          Desktop.getDesktop().browse(new URI(url));
        } catch (Throwable e) {
          Main.showErrorDialog("Error launching browser with URL " + url, e);
        }
      }
    });
    _lastMatchButton.setEnabled(false);
    add(_lastMatchButton, "skip,wrap,span");

  }

  public void updateCurrentMatchUi() {
    HearthstoneMatch match = HearthstoneAnalyser.hsMatch();
    updateMatchClassSelectorsIfSet(match);
    if (_currentMatchEnabled)
      _currentMatchLabel.setText(match.mode() + " Match - " + " Turn " + match.numTurns());
    else
      _currentMatchLabel.setText("Waiting for next match to start ...");
    _currentOpponentNameField.setText(match.opponentName());

    _currentOpponentClassSelect.setSelectedIndex(_getClassOptionIndex(match.opponentClass()));
    _currentYourClassSelector.setSelectedIndex(_getClassOptionIndex(match.userClass()));

    _currentGameCoinField.setSelected(match.coin());
    _currentNotesField.setText(match.notes());
    // last match
    if (_lastMatch != null && _lastMatch.mode() != null) {
      if (_lastMatch.result() != null) {
        String tooltip = (_lastMatch.mode().equals("Arena") ? "View current arena run on"
            : "Edit the previous match") + " on HearthStats.net";
        _lastMatchButton.setToolTipText(tooltip);
        _lastMatchButton.setText(_lastMatch.toString());
        _lastMatchButton.setEnabled(true);
      }
    }
  }

  public void setCurrentMatchEnabled(Boolean enabled) {
    _currentMatchEnabled = enabled;
    _currentYourClassSelector.setEnabled(enabled);
    _currentOpponentClassSelect.setEnabled(enabled);
    _currentGameCoinField.setEnabled(enabled);
    _currentOpponentNameField.setEnabled(enabled);
    _currentNotesField.setEnabled(enabled);
  }

  private int _getClassOptionIndex(String cName) {
    for (int i = 0; i < Constants.hsClassOptions.length; i++) {
      if (Constants.hsClassOptions[i].equals(cName)) {
        return i;
      }
    }
    return 0;
  }

  public void resetMatchClassSelectors() {
    _currentYourClassSelector.setSelectedIndex(0);
    _currentOpponentClassSelect.setSelectedIndex(0);
  }

  public void updateMatchClassSelectorsIfSet(HearthstoneMatch hsMatch) {
    if (_currentYourClassSelector.getSelectedIndex() > 0) {
      hsMatch.userClass_$eq(Constants.hsClassOptions[_currentYourClassSelector.getSelectedIndex()]);
    }
    if (_currentOpponentClassSelect.getSelectedIndex() > 0) {
      hsMatch.opponentClass_$eq(Constants.hsClassOptions[_currentOpponentClassSelect
          .getSelectedIndex()]);
    }
  }
}

File: uploader/src/main/scala/net/hearthstats/ui/MatchPanel.scala
package net.hearthstats.ui

import java.awt.{Color, Desktop, Dimension}
import java.awt.event.{ActionEvent, ActionListener, KeyAdapter, KeyEvent}
import java.net.URI
import javax.swing.BorderFactory.{createCompoundBorder, createEmptyBorder, createMatteBorder}
import javax.swing.{JButton, JCheckBox, JComboBox, JLabel, JPanel, JTextArea, JTextField}
import javax.swing.event.{ChangeEvent, ChangeListener}

import net.hearthstats.{Constants, HearthstoneMatch, Main}
import net.hearthstats.analysis.HearthstoneAnalyser
import net.hearthstats.util.Translations.t
import net.miginfocom.swing.MigLayout

class MatchPanel extends JPanel {

  private val _currentOpponentNameField = new JTextField
  private val _currentMatchLabel = new JLabel
  private val _currentGameCoinField = new JCheckBox(t("match.coin"))
  private val _currentNotesField = new JTextArea()
  private val _lastMatchButton = new JButton("[n/a]")

  var lastMatch: HearthstoneMatch = _

  private var _currentMatchEnabled = false

  setLayout(new MigLayout)
  add(new JLabel(" "), "wrap")
  add(_currentMatchLabel, "skip,span,wrap")
  add(new JLabel(" "), "wrap")

  val localizedClassOptions = Array.ofDim[String](Constants.hsClassOptions.length)
  localizedClassOptions(0) = "- " + t("undetected") + " -"
  for (i <- 1 until localizedClassOptions.length) localizedClassOptions(i) = t(Constants.hsClassOptions(i))

  private val _currentOpponentClassSelect = new JComboBox(localizedClassOptions)
  private val _currentYourClassSelector = new JComboBox(localizedClassOptions)

  add(new JLabel(t("match.label.your_class") + " "), "skip,right")
  add(_currentYourClassSelector, "wrap")
  add(new JLabel(t("match.label.opponents_class") + " "), "skip,right")
  add(_currentOpponentClassSelect, "wrap")
  add(new JLabel("Opponent's Name: "), "skip,right")

  _currentOpponentNameField.setMinimumSize(new Dimension(100, 1))
  _currentOpponentNameField.addKeyListener(new KeyAdapter {
    override def keyReleased(e: KeyEvent) {
      HearthstoneAnalyser.hsMatch.opponentName = _currentOpponentNameField.getText.replaceAll("(\r\n|\n)", "<br/>")
    }
  })

  add(_currentOpponentNameField, "wrap")
  add(new JLabel(t("match.label.coin") + " "), "skip,right")
  _currentGameCoinField.setSelected(false)
  _currentGameCoinField.addChangeListener(new ChangeListener {
    def stateChanged(e: ChangeEvent) {
      HearthstoneAnalyser.hsMatch.coin = _currentGameCoinField.isSelected
    }
  })

  add(_currentGameCoinField, "wrap")
  add(new JLabel(t("match.label.notes") + " "), "skip,wrap")

  _currentNotesField.setBorder(createCompoundBorder(createMatteBorder(1, 1, 1, 1, Color.black), createEmptyBorder(3, 6, 3, 6)))
  _currentNotesField.setMinimumSize(new Dimension(350, 150))
  _currentNotesField.setBackground(Color.WHITE)
  _currentNotesField.addKeyListener(new KeyAdapter {
    override def keyReleased(e: KeyEvent) {
      HearthstoneAnalyser.hsMatch.notes = _currentNotesField.getText
    }
  })

  add(_currentNotesField, "skip,span")
  add(new JLabel(" "), "wrap")
  add(new JLabel(t("match.label.previous_match") + " "), "skip,wrap")
  _lastMatchButton.addActionListener(new ActionListener {
    override def actionPerformed(arg0: ActionEvent) {
      val url = if ("Arena" == lastMatch.mode) "http://hearthstats.net/arenas/new" else lastMatch.editUrl
      try Desktop.getDesktop.browse(new URI(url))
      catch {
        case e: Throwable => Main.showErrorDialog("Error launching browser with URL " + url, e)
      }
    }
  })

  _lastMatchButton.setEnabled(false)
  add(_lastMatchButton, "skip,wrap,span")

  def updateCurrentMatchUi() {
    val currentMatch = HearthstoneAnalyser.hsMatch
    updateMatchClassSelectorsIfSet(currentMatch)
    if (_currentMatchEnabled) _currentMatchLabel.setText(currentMatch.mode + " Match - " + " Turn " + currentMatch.numTurns) else _currentMatchLabel.setText("Waiting for next match to start ...")
    _currentOpponentNameField.setText(currentMatch.opponentName)
    _currentOpponentClassSelect.setSelectedIndex(_getClassOptionIndex(currentMatch.opponentClass))
    _currentYourClassSelector.setSelectedIndex(_getClassOptionIndex(currentMatch.userClass))
    _currentGameCoinField.setSelected(currentMatch.coin)
    _currentNotesField.setText(currentMatch.notes)
    if (lastMatch != null && lastMatch.mode != null) {
      if (lastMatch.result != null) {
        val tooltip = (if (lastMatch.mode == "Arena") "View current arena run on" else "Edit the previous match") +
          " on HearthStats.net"
        _lastMatchButton.setToolTipText(tooltip)
        _lastMatchButton.setText(lastMatch.toString)
        _lastMatchButton.setEnabled(true)
      }
    }
  }

  def setCurrentMatchEnabled(enabled: Boolean) {
    _currentMatchEnabled = enabled
    _currentYourClassSelector.setEnabled(enabled)
    _currentOpponentClassSelect.setEnabled(enabled)
    _currentGameCoinField.setEnabled(enabled)
    _currentOpponentNameField.setEnabled(enabled)
    _currentNotesField.setEnabled(enabled)
  }

  private def _getClassOptionIndex(cName: String): Int = {
    (0 until Constants.hsClassOptions.length).find(Constants.hsClassOptions(_) == cName)
      .getOrElse(0)
  }

  def resetMatchClassSelectors() {
    _currentYourClassSelector.setSelectedIndex(0)
    _currentOpponentClassSelect.setSelectedIndex(0)
    _currentGameCoinField.setSelected(false)
  }

  def updateMatchClassSelectorsIfSet(hsMatch: HearthstoneMatch) {
    if (_currentYourClassSelector.getSelectedIndex > 0) {
      hsMatch.userClass = Constants.hsClassOptions(_currentYourClassSelector.getSelectedIndex)
    }
    if (_currentOpponentClassSelect.getSelectedIndex > 0) {
      hsMatch.opponentClass = Constants.hsClassOptions(_currentOpponentClassSelect.getSelectedIndex)
    }
  }
}
